cmd/tendermint/commands/testnet.go

```diff
@@ -93,205 +58,274 @@ Example:
-	}
+			// Gather persistent peer addresses.
+			var (
+				persistentPeers = make([]string, 0)
+				err             error
+			)
+			tpargs := testnetPeerArgs{
+				numValidators:    nValidators,
+				numNonValidators: nNonValidators,
+				peerToPeerPort:   p2pPort,
+				nodeDirPrefix:    nodeDirPrefix,
+				outputDir:        outputDir,
+				hostnames:        hostnames,
+				startingIPAddr:   startingIPAddress,
+				hostnamePrefix:   hostnamePrefix,
+				hostnameSuffix:   hostnameSuffix,
+				randomMonikers:   randomMonikers,
+			}
 
-	// Write genesis file.
-	for i := 0; i < nValidators+nNonValidators; i++ {
-		nodeDir := filepath.Join(outputDir, fmt.Sprintf("%s%d", nodeDirPrefix, i))
-		if err := genDoc.SaveAs(filepath.Join(nodeDir, config.BaseConfig.Genesis)); err != nil {
-			_ = os.RemoveAll(outputDir)
-			return err
-		}
-	}
+			if populatePersistentPeers {
 
-	// Gather persistent peer addresses.
-	var (
-		persistentPeers = make([]string, 0)
-		err             error
-	)
-	if populatePersistentPeers {
-		persistentPeers, err = persistentPeersArray(config)
-		if err != nil {
-			_ = os.RemoveAll(outputDir)
-			return err
-		}
-	}
+				persistentPeers, err = persistentPeersArray(config, tpargs)
+				if err != nil {
+					_ = os.RemoveAll(outputDir)
+					return err
+				}
+			}
 
-	// Overwrite default config.
-	for i := 0; i < nValidators+nNonValidators; i++ {
-		nodeDir := filepath.Join(outputDir, fmt.Sprintf("%s%d", nodeDirPrefix, i))
-		config.SetRoot(nodeDir)
-		config.P2P.AllowDuplicateIP = true
-		if populatePersistentPeers {
-			persistentPeersWithoutSelf := make([]string, 0)
-			for j := 0; j < len(persistentPeers); j++ {
-				if j == i {
-					continue
+			// Overwrite default config.
+			for i := 0; i < nValidators+nNonValidators; i++ {
+				nodeDir := filepath.Join(outputDir, fmt.Sprintf("%s%d", nodeDirPrefix, i))
+				config.SetRoot(nodeDir)
+				config.P2P.AllowDuplicateIP = true
+				if populatePersistentPeers {
+					persistentPeersWithoutSelf := make([]string, 0)
+					for j := 0; j < len(persistentPeers); j++ {
+						if j == i {
+							continue
+						}
+						persistentPeersWithoutSelf = append(persistentPeersWithoutSelf, persistentPeers[j])
+					}
+					config.P2P.PersistentPeers = strings.Join(persistentPeersWithoutSelf, ",")
+				}
+				config.Moniker = tpargs.moniker(i)
+
+				if err := cfg.WriteConfigFile(nodeDir, config); err != nil {
+					return err
 				}
-				persistentPeersWithoutSelf = append(persistentPeersWithoutSelf, persistentPeers[j])
 			}
-			config.P2P.PersistentPeers = strings.Join(persistentPeersWithoutSelf, ",")
-		}
-		config.Moniker = moniker(i)
 
-		if err := cfg.WriteConfigFile(nodeDir, config); err != nil {
-			return err
-		}
+			fmt.Printf("Successfully initialized %v node directories\n", nValidators+nNonValidators)
+			return nil
+		},
 	}
 
-	fmt.Printf("Successfully initialized %v node directories\n", nValidators+nNonValidators)
-	return nil
+	cmd.Flags().IntVar(&nValidators, "v", 4,
```

Comment:
I might change this to declare the `cmd` struct first, then the vars that will be used by the command function, then add the flags to the cmd struct and finally, add the command function to the cmd struct.